Enhance Android Logging with setLogCallback API

yhoyoon opened the pull request:
This change adds the setLogCallback API to the AndroidChipLogging class of the Matter SDK, enhancing its logging capabilities. With this API, developers can customize logging behavior and integrate it with Android's logging system, allowing for more flexible and comprehensive logging.

An example of usage in Kotlin is as follows:

```kt
ChipDeviceController.loadJni()
AndroidChipLogging.setLogCallback { module, priority, message ->
    when (priority) {
        Log.ERROR -> Log.e(module, message)
        Log.INFO -> Log.i(module, message)
        Log.DEBUG -> Log.d(module, message)
    }
}
```

yhoyoon committed:
Enhance Android Logging with setLogCallback API

This change adds the setLogCallback API to the AndroidChipLogging class
of the Matter SDK, enhancing its logging capabilities. With this API,
developers can customize logging behavior and integrate it with
Android's logging system, allowing for more flexible and comprehensive
logging.
An example of usage in Kotlin is as follows:
ChipDeviceController.loadJni()
AndroidChipLogging.setLogCallback { module, priority, message ->
    when (priority) {
        Log.ERROR -> Log.e(module, message)
        Log.INFO -> Log.i(module, message)
        Log.DEBUG -> Log.d(module, message)
    }
}

Signed-off-by: Youngho Yoon <34558998+yhoyoon@users.noreply.github.com>

tcarmelveilleux commented on src/platform/android/AndroidChipPlatform-JNI.cpp:
```diff
@@ -60,6 +60,9 @@ static bool JavaBytesToUUID(JNIEnv * env, jbyteArray value, chip::Ble::ChipBleUU
 namespace {
 JavaVM * sJVM = nullptr;
 JniGlobalReference sAndroidChipPlatformExceptionCls;
+jclass callbackClass;
+jmethodID onLogMessageMethod;
```
Add `s` prefix for module globals (statics)

yhoyoon replied:
I will fix it.

tcarmelveilleux commented on src/platform/android/AndroidChipPlatform-JNI.cpp:
```diff
@@ -60,6 +60,9 @@ static bool JavaBytesToUUID(JNIEnv * env, jbyteArray value, chip::Ble::ChipBleUU
 JniGlobalReference sAndroidChipPlatformExceptionCls;
+jclass callbackClass;
+jmethodID onLogMessageMethod;
+jobject sLogCallback = nullptr;
```
Rename to  `sJavaLogCallbackObject`

yhoyoon replied:
I will rename it to `sJavaLogCallbackObject`.

tcarmelveilleux commented on src/platform/android/AndroidChipPlatform-JNI.cpp:
```diff
@@ -274,6 +277,63 @@ JNI_LOGGING_METHOD(void, setLogFilter)(JNIEnv * env, jclass clazz, jint level)
+
+    char buffer[CHIP_CONFIG_LOG_MESSAGE_MAX_SIZE];
+    vsnprintf(buffer, sizeof(buffer), msg, args);
+    jstring jMsg = env->NewStringUTF(buffer);
```
Handle the case where jMsg is null due to UTF decode errors in buffer. Don't call the callback when the string is null.

yhoyoon replied:
Thank you. I didn't consider that case. I will modify the code to avoid calling the callback if jMsg is nullptr.

tcarmelveilleux reviewed (approved):
Approval conditional on fixing comments

yhoyoon committed:
Add s prefix for global variables

Signed-off-by: Youngho Yoon <34558998+yhoyoon@users.noreply.github.com>

yhoyoon committed:
Do not call onLogMessage callback when msg is null

Handle the case where jMsg is null due to UTF decode errors in buffer.
Don't call the callback when the string is null.

Signed-off-by: Youngho Yoon <34558998+yhoyoon@users.noreply.github.com>

yunhanw-google commented on src/platform/android/AndroidChipPlatform-JNI.cpp:
```diff
@@ -274,6 +277,72 @@ JNI_LOGGING_METHOD(void, setLogFilter)(JNIEnv * env, jclass clazz, jint level)
+
+    env->CallVoidMethod(sJavaLogCallbackObject, sOnLogMessageMethod, jModule, jPriority, jMsg);
+
+    env->DeleteLocalRef(jModule);
```
I have largely removed DeleteLocalRef usage in matter,  I would prefer using JniLocalReferenceScope scope(env) in the beginning of this function; then you can remove these DeleteLocalRef.
for example https://github.com/project-chip/connectedhomeip/blob/3d69583fb644577269ee468e6a28c0f4d6e40612/src/platform/android/DiagnosticDataProviderImpl.cpp#L101C28-L101C33

yhoyoon replied:
Thank you, I didn't know JniLocalReferenceScope already existed. I prefer that style. 👍

yunhanw-google commented on src/platform/android/AndroidChipPlatform-JNI.cpp:
```diff
@@ -274,6 +277,72 @@ JNI_LOGGING_METHOD(void, setLogFilter)(JNIEnv * env, jclass clazz, jint level)
+{
+    using namespace chip::Logging;
+
+    JNIEnv * env = JniReferences::GetInstance().GetEnvForCurrentThread();
```
please check whether env is nullptr or not

yhoyoon replied:
I will add the line below:

```cpp
    VerifyOrReturn(env != nullptr);
```

yunhanw-google commented on src/platform/android/AndroidChipPlatform-JNI.cpp:
```diff
@@ -60,6 +60,9 @@ static bool JavaBytesToUUID(JNIEnv * env, jbyteArray value, chip::Ble::ChipBleUU
 namespace {
 JavaVM * sJVM = nullptr;
 JniGlobalReference sAndroidChipPlatformExceptionCls;
+jclass sCallbackClass;
```
for global refer, please use JniGlobalReference instead of jclass

yhoyoon replied:
Thank you for your advice. I will change the `sCallbackClass` to `callbackClass` local variable and change the type of `sOnLogMessageMethod` from `jobject` to `JniGlobalReference`.

yunhanw-google reviewed (changes requested):
please kindly address comments, thanks

yunhanw-google commented on src/platform/android/AndroidChipPlatform-JNI.cpp:
```diff
@@ -274,6 +277,72 @@ JNI_LOGGING_METHOD(void, setLogFilter)(JNIEnv * env, jclass clazz, jint level)
+    }
+
+    jint jPriority  = static_cast<jint>(priority);
+    jstring jModule = env->NewStringUTF(module);
```
please don't use NewStringUTF, and go with CharToStringUTF

https://github.com/project-chip/connectedhomeip/blob/3d69583fb644577269ee468e6a28c0f4d6e40612/src/lib/support/JniReferences.h#L237C16-L237C31

yhoyoon replied:
I will change it to use CharToStringUTF.

yhoyoon committed:
Return logRedirectCallback when env is nullptr

Signed-off-by: Youngho Yoon <34558998+yhoyoon@users.noreply.github.com>

yhoyoon committed:
Use JniLocalReferenceScope instead of DeleteLocalRef

Signed-off-by: Youngho Yoon <34558998+yhoyoon@users.noreply.github.com>

yhoyoon committed:
Change type of global ref to JniGlobalReference

Signed-off-by: Youngho Yoon <34558998+yhoyoon@users.noreply.github.com>

yhoyoon committed:
Use CharToStringUTF instead of NewStringUTF

Signed-off-by: Youngho Yoon <34558998+yhoyoon@users.noreply.github.com>

yhoyoon committed:
Use VerifyOrReturn for CharToStringUTF

Signed-off-by: Youngho Yoon <34558998+yhoyoon@users.noreply.github.com>
